sel/src/Sel/PublicKey/Internal/Signature.hs

```diff
@@ -0,0 +1,499 @@
+decodeSecretKeyHexByteString bytes = SecretKey <$> copyHexKey cryptoSignSecretKeyBytes bytes
+
+-- | Encode an t'UnsafeSecretKey' to a hexadecimal encoded
+-- 'StrictByteString' using the default copying encoder.
+--
+-- @since 0.0.3.0
+encodeSecretKeyHexByteString :: UnsafeSecretKey -> StrictByteString
+encodeSecretKeyHexByteString (UnsafeSecretKey (SecretKey secretKeyPtr)) =
+  Base16.extractBase16 . Base16.encodeBase16' $
+    ByteString.fromForeignPtr0
+      (Foreign.castForeignPtr secretKeyPtr)
+      (fromIntegral cryptoSignSecretKeyBytes)
+
+-- | Produce the t'PublicKey' from a t'SecretKey'.
+--
+-- This function may throw a t'PublicKeyExtractionException' if the operation fails.
+publicKey :: SecretKey -> PublicKey
+publicKey (SecretKey secretKeyPtr) = unsafeDupablePerformIO $ do
+  publicKeyPtr <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSignPublicKeyBytes)
+  res <-
+    useM $
+      cryptoSignED25519SkToPk
+        <$> foreignPtr publicKeyPtr
+        <*> foreignPtr secretKeyPtr
+  unless (res == 0) $ throw PublicKeyExtractionException
+  pure $ PublicKey publicKeyPtr
+
+-- | Signal your intent to encode secret key material for transmission
+-- by wrapping a t'SecretKey' in t'UnsafeSecretKey'.
+--
+-- @since 0.0.3.0
+newtype UnsafeSecretKey = UnsafeSecretKey SecretKey
+  deriving newtype
+    ( Eq
+      -- ^ @since 0.0.3.0
+      -- Follows the t'SecretKey' instance.
+    )
+  deriving
+    ( Display
+      -- ^ @since 0.0.3.0
+      -- Hexadecimal-encoded bytes.
+    )
+    via (ShowInstance UnsafeSecretKey)
+
+-- | Hexadecimal-encoded bytes.
+--
+-- @since 0.0.3.0
+instance Show UnsafeSecretKey where
+  show = ByteString.unpackChars . encodeSecretKeyHexByteString
+
+-- | By lexicographical comparison of pointer contents.
+--
+-- ⚠️ Vulnerable to timing attacks!
+--
+-- @since 0.0.3.0
+instance Ord UnsafeSecretKey where
+  a `compare` b =
+    foreignPtrOrd (coerce a) (coerce b) cryptoSignSecretKeyBytes
+
+-- | A signing key pair, comprising a t'PublicKey' and a t'SecretKey'.
+--
+-- @since 0.0.3.0
+data KeyPair = KeyPair {public :: PublicKey, secret :: SecretKey}
+  deriving stock
+    ( Show
+      -- ^ @since 0.0.3.0
+      -- Follows the instances for t'PublicKey' and t'SecretKey', respectively.
+      --
+      -- In particular, the secret key will be shown as @[REDACTED]@.
+    , Eq
+      -- ^ @since 0.0.3.0
+      -- Follows the instances for t'PublicKey' and t'SecretKey', respectively.
+    )
+  deriving
+    ( Display
+      -- ^ @since 0.0.3.0
+      -- Follows the instances for t'PublicKey' and t'SecretKey', respectively.
+      --
+      -- In particular, the secret key will be displayed as @[REDACTED]@.
+    )
+    via (ShowInstance KeyPair)
+
+-- | By lexicographical comparison of key pointer contents.
+--
+-- ⚠️ Vulnerable to timing attacks!
+--
+-- @since 0.0.3.0
+instance Ord KeyPair where
+  compare kp1 kp2 =
+    compare kp1.public kp2.public
+      <> comparing UnsafeSecretKey kp1.secret kp2.secret
```

Comment:
Since it's a default instance, people ought to be able to use it (even indirectly) without having to think twice about timing attacks. Is it possible to make use of the time-safe functions?

Reply:
https://libsodium.gitbook.io/doc/helpers#constant-time-test-for-equality is what we have in the existing bindings for constant-time comparisons, but that only gives us equal/not-equal.

It looks like I _might_ be able to use https://libsodium.gitbook.io/doc/helpers#comparing-large-numbers, but there's no binding for that one in `libsodium-bindings` at the moment. Happy to add the binding and use that instead of the `UnsafeSecretKey` if that'd work for you.

Reply:
I think that it would work for me yes. :)